feat(internal/rust): add semantic analysis

affo opened the pull request:
Closes #1972 
Closes #1955  

### Done checklist
- [ ] docs/SPEC.md updated
- [x] Test cases written

jlapacik commented on libflux/src/semantic/nodes.rs:
```diff
@@ -0,0 +1,567 @@
+    Ret(ReturnStatement<'a>),
+    Test(TestStatement<'a>),
+    Built(BuiltinStatement<'a>),
+}
```
These variant names will probably have to be updated to reflect the AST.

jlapacik commented on libflux/src/semantic/nodes.rs:
```diff
@@ -0,0 +1,567 @@
+#[derive(Debug, PartialEq, Clone)]
+pub struct TypeInformation {
+    pub mono_type: types::MonoType,
+}
```
Is there a reason for wrapping the monotype vs attaching the monotype to the node directly?

affo replied:
It was mostly to implement `update` once and forall.

Do you preder to remove the method and attache only the monotype?

jlapacik commented on libflux/src/semantic/nodes.rs:
```diff
@@ -0,0 +1,567 @@
+#[derive(Debug, PartialEq, Clone)]
+pub struct StringExpression<'a> {
+    pub loc: Location<'a>,
+    pub type_info: TypeInformation,
```
Could this field just be of type `MonoType`?

affo replied:
see https://github.com/influxdata/flux/pull/2027#discussion_r336613955

jlapacik commented on libflux/src/semantic/nodes.rs:
```diff
@@ -0,0 +1,567 @@
+    pub type_info: TypeInformation,
+
+    pub callee: Expression<'a>,
+    pub arguments: ObjectExpression<'a>,
```
I don't think this should be an object expression. This says that every function implicitly takes a single object as a parameter. But this isn't consistent with the ability of a function to specify default or pipe arguments. Nor is it consistent with extensible records which is what our objects are now.

affo replied:
Mmmm, I guess the ability of a function to specify default or pipe arguments should be in `FunctionExpr`, and that's where it is.

`CallExpr`, instead, is only to specify how a function is called.

> Nor is it consistent with extensible records which is what our objects are now

For this, how should I change the representation then? In my last commit I opted for `Vec<Property>`.

jlapacik commented on libflux/src/semantic/nodes.rs:
```diff
@@ -0,0 +1,567 @@
+    pub loc: Location<'a>,
+
+    pub key: PropertyKey<'a>,
+    pub value: Option<Expression<'a>>,
```
Why is this optional?

affo replied:
There can be cases in which the objects do not specify values. For example:

```
{a, b}
```

Is equivalent to

```
{a: a, b: b}
```

Now I see your point. I guess I could remove the  optional expression and assign an IdentifierExpr in that case.

affo replied:
I have done that in the last commit.

jlapacik commented on libflux/src/semantic/analyze.rs:
```diff
@@ -0,0 +1,2688 @@
+    // TestFresher mocks the fresher so that we don't have to test Tvar ids.
+    // That is not straightforward because the IncrementingFresher is used at different depths in
+    // the semantic graph.
+    struct TestFresher {}
```
I believe the derivative crate will allow you to do something like this:
```Rust
#[derive(Derivative)]
#[derivative(PartialEq)]
pub struct Expression {
    ....
    #[derivative(PartialEq="ignore")]
    pub mono_type: MonoType,
}
```
so that you can ignore the monotype. Do we want to do that? We could ignore the source location as well.

On an "unrelated" note, I now see why you made `Fresher` a trait.

affo replied:
Mmmmm, do we really want to make equality ignore the source location?

Two  identical identifiers, for instance, in two different positions are not the same identifier.

Even though I like this Rust directive, I wouldn't like to change the nodes' behavior for tests.

I am waiting for your feedback on this.
(I would really like to have some like Go's comparison options)

affo replied:
Thinking more about it, the type can be excluded from the comparison, but I'd avoid to exclude the location for the reason above.

We could say that, in general, no two nodes in the same place should have a different type. So, we can safely exclude the type.

affo commented on libflux/src/semantic/analyze.rs:
```diff
@@ -0,0 +1,2688 @@
+        .body
+        .iter()
+        .map(|s| analyze_statement(s, fresher))
+        .collect::<Result<Vec<Statement>>>()?;
```
🙏

affo commented on libflux/src/semantic/analyze.rs:
```diff
@@ -0,0 +1,2688 @@
+use std::result;
+
+type SemanticError = String;
+type Result<T> = result::Result<T, SemanticError>;
```
🙏

affo commented on libflux/src/semantic/analyze.rs:
```diff
@@ -0,0 +1,2688 @@
+        ast::Expression::Time(lit) => Ok(Expression::Time(analyze_time_literal(&lit, fresher)?)),
+        ast::Expression::PipeLit(_) => Err("a pipe literal may only be used as a default value for an argument in a function definition".to_string()),
+        // This should not happen, but it could for BadExpression.
+        _ => Err("unsupported expression in semantic analysis".to_string())
```
@jlapacik this shouldn't.

The ast shouldn't contain any error node at this point.

The check should be run before passing the ast to `analyze`.

jlapacik replied:
I agree the AST shouldn't contain any error nodes. So if we get an AST with a BadExpression node, we should error. The reason I say this should explicitly match on BadExpression, is that if a new node is added to the AST, then this won't compile.

affo commented on libflux/src/semantic/nodes.rs:
```diff
@@ -0,0 +1,567 @@
+use chrono::FixedOffset;
+use std::vec::Vec;
+
+pub type Location<'a> = &'a ast::SourceLocation;
```
Yep, the idea is to point to that location and avoid cloning it. It should be useless.

affo commented on libflux/src/semantic/types.rs:
```diff
@@ -119,14 +119,38 @@ impl fmt::Display for MonoType {
+// Fresher is used to get fresh type variables.
+pub trait Fresher {
+    fn get(&mut self) -> Tvar;
+}
```
I maintained the trait implementation for mocking the fresher.

affo commented on libflux/src/semantic/nodes.rs:
```diff
@@ -0,0 +1,567 @@
+    pub loc: Location<'a>,
+
+    pub key: PropertyKey<'a>,
+    pub value: Expression<'a>,
```
I have updated this not to be optional (I realized that that was useless as you noted).

affo commented on libflux/src/semantic/nodes.rs:
```diff
@@ -0,0 +1,567 @@
+    pub type_info: TypeInformation,
+
+    pub callee: Expression<'a>,
+    pub arguments: Vec<Property<'a>>,
```
Now Arguments are a list of properties

affo commented:
@jlapacik I passed to a more OO model in order to simplify method signatures and use a fresher in the `Analyzer`.

That  way it can be mocked easily. I fixed the `ObjectExpr` thing in `CallExpr` and fixed the naming coherently with the changes in `ast`.

jlapacik commented on libflux/src/semantic/types.rs:
```diff
@@ -119,14 +119,28 @@ impl fmt::Display for MonoType {
+        Fresher(0)
+    }
+
+    pub fn get(&mut self) -> Tvar {
```
Could you name this `fresh` instead of `get`? It's just a little more explicit about what is actually happening. I usually don't expect a `get` method to mutate the object.

jlapacik commented on libflux/src/semantic/analyze.rs:
```diff
@@ -0,0 +1,2604 @@
+                Ok(Statement::Variable(self.analyze_variable_assignment(&s)?))
+            }
+            // This should not happen, but it could for BadStatement.
+            _ => Err("unsupported statement in semantic analysis".to_string()),
```
Similar to `analyze_expression` but the only option here is that we get a `BadStatement` node right? If that's the case, can we explicitly match on `BadStatement`?

To be clear, I think we should still error, I just prefer the stronger compile time guarantees that matching on all cases gives us.

affo replied:
Yeah, wise 👍

jlapacik commented on libflux/src/semantic/analyze.rs:
```diff
@@ -0,0 +1,2604 @@
+        // TODO(affo): I'd prefer these checks to be in ast.Check().
+        if expr.arguments.len() > 1 {
+            return Err("arguments are more than one object expression".to_string());
+        }
```
This can be removed now.

jlapacik replied:
Actually never mind. I misread this. Although I would definitely prefer that this be changed in the AST. But not necessary for this PR.

jlapacik commented on libflux/src/semantic/analyze.rs:
```diff
@@ -0,0 +1,2580 @@
+type Result<T> = result::Result<T, SemanticError>;
+
+// analyze analyzes an AST package node and returns its semantic analysis.
+// The function requires a Fresher to instantiate fresh type veraibles for Expressions.
```
typo: variables

jlapacik commented on libflux/src/semantic/analyze.rs:
```diff
@@ -0,0 +1,2580 @@
+
+// analyze analyzes an AST package node and returns its semantic analysis.
+// The function requires a Fresher to instantiate fresh type veraibles for Expressions.
+// The function explicitely moves the ast::Package because it adds information to it.
```
typo: explicitly

jlapacik commented on libflux/src/semantic/analyze.rs:
```diff
@@ -0,0 +1,2580 @@
+// The function explicitely moves the ast::Package because it adds information to it.
+// We follow here the principle that every compilation step should be isolated and should add meaning
+// to the previous one. In other terms, once one analyzes an AST he should not use it anymore.
+// If one wants to do so, he should explicitely pkg.clone() and incur consciously in the memory
```
typo: explicitly

jlapacik commented on libflux/src/semantic/analyze.rs:
```diff
@@ -0,0 +1,2580 @@
+
+fn type_info(fresher: &mut Fresher) -> MonoType {
+    MonoType::Var(fresher.fresh())
+}
```
I think we can replace calls to `type_info` now with `MonoType::Var(fresher.fresh())` to be more explicit.

jlapacik commented on libflux/src/semantic/nodes.rs:
```diff
@@ -0,0 +1,576 @@
+    pub loc: ast::SourceLocation,
+
+    pub expression: Expression,
+}
```
TextPart and InterpolatedPart aren't considered expressions on their own and so they aren't given types. However InterpolatedPart contains an expression. This expression will be constrained to be of type `string` initially. Eventually though it will just be constrained to be of kind `StringFormattable`(insert better name here).

affo replied:
Yeah, my assumption was that the expression in the interpolated part will be constrained to work, but in the end it would be turned into a string.

affo commented on libflux/src/semantic/analyze.rs:
```diff
@@ -0,0 +1,2604 @@
+type SemanticError = String;
+type Result<T> = result::Result<T, SemanticError>;
+
+pub struct Analyzer {
```
This is no more there, sorry 😅

affo commented on libflux/src/semantic/types.rs:
```diff
@@ -119,14 +119,28 @@ impl fmt::Display for MonoType {
 }
 
+// Fresher returns a fresh type variable with incrementing id.
+pub struct Fresher(i64);
```
yep, it needs to because it is used in `analyze`.

I could move it to analyze and make it private because I guess it wouldn't be used anywhere else.

jlapacik replied:
Sorry I missed this. `analyze` will need to take a `Fresher` as input and so it should be public so that the same `Fresher` can be used when analyzing multiple packages.

affo replied:
Yeah. I guessed that was the reason. I can revert the last commit then.

affo commented:
@jlapacik I fixed typos and addressed the `typ` minor change.

@nathanielc 
In the last commit I made the `Fresher` private by moving it to `analyze`. This implies that `fn analyze` cannot take a `Fresher` in. Could this be a problem @jlapacik ? 🤔 

I can revert that last commit if there is an explicit need to take a fresher in.

affo commented:
@jlapacik @nathanielc rebased to fix conflicts and reverted the last commit. Is this good to go?
